{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "# type the gene's name in the gene_id_list here!!\n",
    "gene_id_list = [\n",
    "    \"OR52M1\", \"OR51B5\", \"OR52N4\", \"OR11H1\", \"OR4K15\", \"OR6S1\", \"OR5AP2\", \"OR5H1\", \"OR13C2\", \"OR2K2\", \"OR8G5\", \"OR1L3\", \"OR2T27\"]\n",
    "saved_list = []\n",
    "\n",
    "\n",
    "def fetch(jsondata, url=\"https://gnomad.broadinstitute.org/api\"):\n",
    "    # The server gives a generic error message if the content type isn't\n",
    "    # explicitly set\n",
    "    headers = {\"Content-Type\": \"application/json\"}\n",
    "    response = requests.post(url, json=jsondata, headers=headers)\n",
    "    json = response.json()\n",
    "    if \"errors\" in json:\n",
    "        raise Exception(json[\"errors\"])\n",
    "    return json\n",
    "\n",
    "\n",
    "def get_variant_list(gene_id, dataset=\"gnomad_r2_1\"):\n",
    "    # Note that this is GraphQL, not JSON.\n",
    "    fmt_graphql = \"\"\"\n",
    "    {\n",
    "        gene(gene_name: \"%s\") {\n",
    "            variants(dataset: %s) {\n",
    "                consequence\n",
    "                pos\n",
    "                rsid\n",
    "                ref\n",
    "                alt\n",
    "                flags\n",
    "                lof\n",
    "                consequence_in_canonical_transcript\n",
    "                gene_symbol\n",
    "                chrom\n",
    "                hgvsc\n",
    "                reference_genome\n",
    "                exome {\n",
    "                  ac\n",
    "                  ac_hemi\n",
    "                  ac_hom\n",
    "                  an\n",
    "                  af\n",
    "                }  \n",
    "                lof_filter\n",
    "                genome {\n",
    "                  ac\n",
    "                  ac_hemi\n",
    "                  ac_hom\n",
    "                  an\n",
    "                  af\n",
    "                }\n",
    "                lof_flags\n",
    "                hgvsc\n",
    "                hgvsp\n",
    "                reference_genome\n",
    "                variant_id: variantId\n",
    "            }\n",
    "        }\n",
    "      }\n",
    "    \"\"\"\n",
    "    # This part will be JSON encoded, but with the GraphQL part left as a\n",
    "    # glob of text.\n",
    "    req_variantlist = {\n",
    "        \"query\": fmt_graphql % (gene_id, dataset),\n",
    "        \"variables\": {}\n",
    "    }\n",
    "    response = fetch(req_variantlist)\n",
    "    return response[\"data\"][\"gene\"][\"variants\"]\n",
    "\n",
    "\n",
    "# add the gene_id at the last column\n",
    "def make_df(gene_id):\n",
    "    li = get_variant_list(gene_id)\n",
    "    df = pd.DataFrame(li)\n",
    "    return df\n",
    "\n",
    "\n",
    "# get_header\n",
    "header = [\n",
    "    'Name',\n",
    "    'Chromosome',\n",
    "    'Position',\n",
    "    'rsID',\n",
    "    'Reference',\n",
    "    'Alternate',\n",
    "    'Protein Consequence',\n",
    "    'Transcript Consequence',\n",
    "    'Annotation',\n",
    "    'Flags',\n",
    "    'Allele Count',\n",
    "    'Allele Number',\n",
    "    'Allele Frequency',\n",
    "    'Homozygote Count',\n",
    "    'Hemizygote Count',\n",
    "    'Variant ID'\n",
    "]\n",
    "\n",
    "\n",
    "def make_new_df(old_df):\n",
    "    df = old_df[old_df['exome'].notnull()]\n",
    "    reframe_data = []\n",
    "    for row in range(len(df)):\n",
    "        newarray = []\n",
    "        newarray.append(df.iloc[row].get('gene_symbol')) # 1\n",
    "        newarray.append(df.iloc[row].get('chrom')) # 2\n",
    "        newarray.append(df.iloc[row].get('pos')) # 3\n",
    "        newarray.append(df.iloc[row].get('rsid'))  # 4\n",
    "        newarray.append(df.iloc[row].get('ref')) # 5\n",
    "        newarray.append(df.iloc[row].get('alt')) # 6\n",
    "        newarray.append(df.iloc[row].get('hgvsp')) # 7\n",
    "        newarray.append(df.iloc[row].get('hgvsc')) # 8\n",
    "        newarray.append(df.iloc[row].get('consequence')) # 9\n",
    "        if df.iloc[row].get('flags'):\n",
    "            newarray.append(df.iloc[row].get('flags')[0]) # 10\n",
    "        else:\n",
    "            newarray.append('')\n",
    "        newarray.append(df.iloc[row].get('exome').get('ac')) # 11\n",
    "        newarray.append(df.iloc[row].get('exome').get('an')) # 12\n",
    "        newarray.append(df.iloc[row].get('exome').get('af')) # 13\n",
    "        newarray.append(df.iloc[row].get('exome').get('ac_hom')) # 14\n",
    "        newarray.append(df.iloc[row].get('exome').get('ac_hemi')) # 15\n",
    "        newarray.append(df.iloc[row].get('variant_id')) # 16\n",
    "        reframe_data.append(newarray)\n",
    "    #add reframe_data and header\n",
    "    new_df = pd.DataFrame(np.array(reframe_data), columns=header)\n",
    "    return new_df\n",
    "\n",
    "# generate and download csv file of each gene, saved to current path\n",
    "def generate_csv(gene_id_list):\n",
    "    for gene_id in gene_id_list:\n",
    "        try:\n",
    "            make_new_df(make_df(gene_id)).to_csv(gene_id + '.csv',index=False)    \n",
    "            saved_list.append(gene_id)\n",
    "            print('\\x1b[6;30;42m' + 'Saved: ' + gene_id + '\\x1b[0m')\n",
    "        except:\n",
    "            print('\\x1b[6;30;41m' + 'Error: ' + gene_id + '\\x1b[0m')\n",
    "            pass\n",
    "\n",
    "# download the combined csv file to current path, named as \"output.csv\"\n",
    "def combine_csv():\n",
    "    generate_csv(gene_id_list)\n",
    "    combined_csv = pd.read_csv(saved_list[0] + '.csv')\n",
    "    combined_csv.to_csv('output.csv', encoding=\"utf_8_sig\", index=False)\n",
    "    for gene_id in saved_list:\n",
    "        combined_csv = pd.read_csv(gene_id + '.csv')\n",
    "        combined_csv.to_csv('output.csv',encoding=\"utf_8_sig\",index=False,header=False,mode='a+')\n",
    "    print('Combining file...\\n\\x1b[6;30;42mConbined File to output.csv\\x1b[6;30;41m')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[6;30;42mSaved: OR52M1\u001b[0m\n",
      "\u001b[6;30;42mSaved: OR51B5\u001b[0m\n",
      "\u001b[6;30;42mSaved: OR52N4\u001b[0m\n",
      "\u001b[6;30;42mSaved: OR11H1\u001b[0m\n",
      "\u001b[6;30;42mSaved: OR4K15\u001b[0m\n",
      "\u001b[6;30;42mSaved: OR6S1\u001b[0m\n",
      "\u001b[6;30;42mSaved: OR5AP2\u001b[0m\n",
      "\u001b[6;30;42mSaved: OR5H1\u001b[0m\n",
      "\u001b[6;30;41mError: OR13C2\u001b[0m\n",
      "\u001b[6;30;42mSaved: OR2K2\u001b[0m\n",
      "\u001b[6;30;42mSaved: OR8G5\u001b[0m\n",
      "\u001b[6;30;42mSaved: OR1L3\u001b[0m\n",
      "\u001b[6;30;42mSaved: OR2T27\u001b[0m\n",
      "Combining file...\n",
      "\u001b[6;30;42mConbined File to output.csv\u001b[6;30;41m\n"
     ]
    }
   ],
   "source": [
    "combine_csv()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
